File: rotators/aws-lambda/src/mysql.ts
import mysql from "mysql2/promise";

export interface MysqlRotateParams {
    type: "mysql";
    host: string;
    port: number;
    database: string;

    managingUser: MysqlUser;
    rotateUser: MysqlUser;
}

interface MysqlUser {
    username: string;
    password: string;
    newPassword?: string;
}

export default async function rotate(event: MysqlRotateParams) {
    let statusCode = 200;
    let message = "ok";

    const { database, host, port, managingUser, rotateUser } = event;
    const { username: managingUsername, password: managingPassword } = managingUser;
    const { username: rotateUsername, newPassword: newPassword } = rotateUser;

    if (!newPassword) {
        return {
            code: 400,
            message: "No new password provided",
        }
    }

    try {
        const connection = await mysql.createConnection({
            host: host,
            port: port,
            database: database,
            user: managingUsername,
            password: managingPassword,
        });

        const query = "ALTER USER ? IDENTIFIED BY ?";
        await connection.query(query, [rotateUsername, newPassword]);
    } catch (error) {
        statusCode = 400;
        message = `Error rotating user '${rotateUsername}': ${error}`;
    }

    return {
        code: statusCode,
        message: message,
    }
}

File: rotators/aws-lambda/mysql/mysql.go
package mysql

import (
	"context"
	"fmt"
	"net"

	"github.com/go-sql-driver/mysql"
)
import "database/sql"

type MysqlRotateParams struct {
	Host     string `json:"host"`
	Port     int    `json:"port"`
	Database string `json:"database"`

	ManagingUser MysqlUser `json:"managingUser"`
	RotateUser   MysqlUser `json:"rotateUser"`
}

type MysqlUser struct {
	Username    string  `json:"username"`
	Password    string  `json:"password"`
	NewPassword *string `json:"newPassword"`
}

func Rotate(ctx context.Context, request MysqlRotateParams) error {
	if request.RotateUser.NewPassword == nil {
		return fmt.Errorf("no password provided")
	}

	db, err := sql.Open("mysql", (&mysql.Config{
		Net:               "tcp",
		Addr:              net.JoinHostPort(request.Host, fmt.Sprintf("%d", request.Port)),
		DBName:            request.Database,
		User:              request.ManagingUser.Username,
		Passwd:            request.ManagingUser.Password,
		InterpolateParams: true,
	}).FormatDSN())
	if err != nil {
		return err
	}
	defer db.Close()

	if err = db.PingContext(ctx); err != nil {
		return fmt.Errorf("connecting to database: %w", err)
	}

	_, err = db.ExecContext(ctx, `ALTER USER ? IDENTIFIED BY ?`, request.RotateUser.Username, request.RotateUser.NewPassword)
	if err != nil {
		return fmt.Errorf("error rotating user %q: %w", request.RotateUser.Username, err)
	}
	return nil
}
